WalletWasabi/WabiSabi/Client/CredentialDependencies/DependencyGraph.cs

```diff
@@ -0,0 +1,350 @@
+				{
+					var largestIsSink = largestMagnitudeNode!.InitialBalance(credentialType).CompareTo(smallMagnitudeNodes!.First().InitialBalance(credentialType)) == -1;
+					var maxCount = K - (largestIsSink ? InDegree(largestMagnitudeNode, credentialType) : OutDegree(largestMagnitudeNode, credentialType));
+
+					if (Math.Abs(Balance(largestMagnitudeNode, credentialType)) > Math.Abs(smallMagnitudeNodes!.Sum(x => Balance(x, credentialType))))
+					{
+						// When we are draining a positive valued node into
+						// multiple negative nodes and we can't drain it
+						// completely, we need to leave an edge unused for the
+						// remaining amount.
+						// The corresponding condition can't actually happen for fan-in
+						// because the negative balance of the last loop
+						// iteration can't exceed the the remaining positive
+						// elements, their total sum must be positive as checked
+						// in the constructor.
+						maxCount--;
+					}
+
+					// Reduce the number of small magnitude nodes to the number
+					// of edges available for use in the largest magnitude node
+					smallMagnitudeNodes = LimitCount(smallMagnitudeNodes!, maxCount, credentialType);
+
+					// After draining either the last small magnitude node or
+					// the largest magnitude node could still have a non-zero
+					// value.
+					DrainTerminal(largestMagnitudeNode!, smallMagnitudeNodes!, credentialType);
+				}
+
+				// at this point the sub-graph of credentialType edges should be
+				// a planar DAG with the AssertResolvedGraphInvariants() holding for
+				// that particular type.
+			}
+
+			// at this point the entire graoh should be a DAG with labeled
+			// edges that can be partitioned into NumTypes different planar
+			// DAGs, and the invariants should hold for all of these.
+		}
+
+		// Find the largest negative or positive balance node for the given
+		// credential type, and one or more smaller nodes with a combined total
+		// magnitude exceeding that of the largest magnitude node when possible.
+		public bool SelectNodesToDischarge(CredentialType credentialType, out RequestNode? largestMagnitudeNode, out IEnumerable<RequestNode>? smallMagnitudeNodes)
+		{
+			// Order the nodes of the graph based on their balances
+			var ordered = VerticesByBalance(credentialType);
+			ImmutableArray<RequestNode> positive = ordered.ThenBy(x => OutDegree(x, credentialType)).Where(v => Balance(v, credentialType) > 0).ToImmutableArray();
+			ImmutableArray<RequestNode> negative = Enumerable.Reverse(ordered.ThenByDescending(x => InDegree(x, credentialType)).Where(v => Balance(v, credentialType) < 0)).ToImmutableArray();
+
+			if (negative.Length == 0)
+			{
+				largestMagnitudeNode = null;
+				smallMagnitudeNodes = null;
+				return false;
+			}
+
+			var nPositive = 1;
+			var nNegative = 1;
+
+			IEnumerable<RequestNode> posCandidates() => positive.Take(nPositive);
```

Comment:
![image](https://user-images.githubusercontent.com/9844978/117290740-73f4f100-ae6e-11eb-9b26-42f9a57db15f.png)

Reply:
i'm not sure what the red lines are indicating, in my editor that's how compilation errors are displayed, but this code builds & passes tests...